### ISSUE TITLE
Aggregated measurements have higher values in Materialized View than in the table which it built on

### ISSUE THREAD

**Issue by mkocztorz** [ClickHouse/ClickHouse#25853]
**Describe the unexpected behaviour**

When connecting to one node and performing inserts on Distributed table the data in Materialized View do not match aggregated values from original table, but only on a shard where the inserts are NOT done.

We have four nodes:
- Shard 1 Replica 1
- Shard 1 Replica 2
- Shard 2 Replica 1
- Shard 2 Replica 2

I have Distributed table `events` designed for writes and reads from `event_shard` which is `ReplicatedSummingMergeTree` engine.
On each node there is a  `Materialized View` called `event_hourly_shard` which holds same data as `event_shard` just preaggregated to full hours.

Tables are created ON CLUSTER.

The event_shard table
---
```
CREATE TABLE IF NOT EXISTS insights.event_shard ON CLUSTER 'clickhouse-main'
(
    event_at     DateTime,
    visit_at     DateTime,
    event_type   String,
    publisher_id UInt32,
    count        UInt64
)
    ENGINE = ReplicatedSummingMergeTree('/clickhouse/tables/{cluster}/{shard}/event', '{replica}')
        PARTITION BY event_at
        PRIMARY KEY (event_at, visit_at, event_type, publisher_id)
        ORDER BY (event_at, visit_at, event_type, publisher_id)
```
Distributed event table (used for inserts)
---
```
CREATE TABLE IF NOT EXISTS insights.event ON CLUSTER 'clickhouse-main'
(
    event_at     DateTime,
    visit_at     DateTime,
    event_type   String,
    publisher_id UInt32,
    count        UInt64
)
    ENGINE = Distributed('clickhouse-main', insights, event_shard, rand())
```
Materialized view for event_shard table (hourly aggregate):
---
```
CREATE MATERIALIZED VIEW IF NOT EXISTS insights.event_hourly_shard
            ON CLUSTER 'clickhouse-main'
            ENGINE = ReplicatedSummingMergeTree('/clickhouse/tables/{cluster}/{shard}/event_hourly', '{replica}')
                PARTITION BY event_at
                PRIMARY KEY (event_at, visit_at, event_type, publisher_id)
                ORDER BY (event_at, visit_at, event_type, publisher_id)
            POPULATE
AS
SELECT toStartOfHour(event_at) as event_at,
       toStartOfHour(visit_at) as visit_at,
       event_type,
       publisher_id,
       sum(count) as count
FROM insights.event_shard
GROUP BY toStartOfHour(event_at) as event_at,
         toStartOfHour(visit_at) as visit_at,
         event_type,
         publisher_id
```
Additional Distributed table on top of event_hourly_shard (not part of the issue, just to show full picture)
---
```
CREATE TABLE IF NOT EXISTS insights.event_hourly ON CLUSTER 'clickhouse-main'
(
    event_at        DateTime,
    visit_at        DateTime,
    event_type      String,
    publisher_id    UInt32,
    count           UInt64
)
    ENGINE = Distributed('clickhouse-main', insights, event_hourly_shard, rand())
```

**How to reproduce**

Insert data to Distributed `event` table only on `Shard 2 Replica 2`. I know it's not optimal.

Then, run query on Shard 1 and Shard 2 (any of 2 replicas)
```
WITH 1 as pid, '2021-06-29' as date, 14 as hour
SELECT 'quarterly' as version, toDate(visit_at), sum(count)
FROM insights.event_shard
where publisher_id = pid
  and toHour(visit_at) = hour
  and toDate(visit_at) = date
group by toDate(visit_at), event_type
ORDER BY toDate(visit_at) desc

UNION ALL

SELECT 'hourly' as version, toDate(visit_at), sum(count)
FROM insights.event_hourly_shard
where publisher_id = pid
  and toHour(visit_at) = hour
  and toDate(visit_at) = date
group by toDate(visit_at)
ORDER BY toDate(visit_at) desc;
```

**Expected behavior**

This is result of above query on Shard 2 (the one the inserts were done) - both counts are equal as expected

table        | date           | count
------------- |-------------| -----
quarterly|2021-06-29|29538570
hourly|2021-06-29|29538570

**Unexpected behavior**

This is result of above query on Shard 1 (where no inserts were done directly) - counts are different

table        | date           | count
------------- |-------------| -----
quarterly|2021-06-29|29548831
hourly|2021-06-29|30467668

What is even more puzzling, the value from materialized view is higher than the value from event_shard table.

I confirmed the values in `event_shard` table are valid. The higher count from `event_hourly_shard` is invalid.

When I modified the code to insert data to `event_shard` directly and distributed the inserts, then both shards behave correctly. This points the issue is somewhere on the `event` Distributed table, while forwarding the inserts to the other shard.

* Which ClickHouse server version to use
`ClickHouse server version 21.5.7.9 (official build).`

**Comment by den-crane** [comment]
```sql
insights.event_shard
...
 PARTITION BY event_at    <<<<---- really ?? Or it just a typo in the issue?
```

Please share `select * from system.settings where changed`.

Distributed table repeats insertion on failure automatically. Though with default settings it should not be a problem.
Check CH logs at the node that you use for insertion.

**Comment by mkocztorz** [comment]
Yes it's partitioned by event_at, but before inserting its grouped to 15 minutes.

Currently we are recreating the cluster. When it's back I will repeat the test and share the changed settings and logs. But as far as I know we were not changing many defaults.

**Comment by mkocztorz** [comment]
Updated config
---
|name|value                        |changed|description                                  |min|max|readonly|type         |
|----|-----------------------------|-------|---------------------------------------------|---|---|--------|-------------|
|use_uncompressed_cache|0                            |1      |Whether to use the cache of uncompressed blocks.|   |   |0       |Bool         |
|load_balancing|in_order                     |1      |Which replicas (among healthy replicas) to preferably send a query to (on the first attempt) for distributed processing.|   |   |0       |LoadBalancing|
|log_queries|1                            |1      |Log requests and write the log to the system table.|   |   |0       |Bool         |
|max_memory_usage|10000000000                  |1      |Maximum memory usage for processing of single query. Zero means unlimited.|   |   |0       |UInt64       |
|max_partitions_per_insert_block|1000                         |1      |Limit maximum number of partitions in single INSERTed block. Zero means unlimited. Throw exception if the block contains too many partitions. This setting is a safety threshold, because using large number of partitions is a common misconception.|   |   |0       |UInt64       |
|format_csv_delimiter|;                            |1      |The character to be considered as a delimiter in CSV data. If setting with a string, a string has to have a length of 1.|   |   |0       |Char         |

Sample log
---
```
2021.07.14 10:00:54.540257 [ 11001 ] {} <Error> insights.insights.event_hourly_shard (ReplicatedMergeTreeRestartingThread): Couldn't start replication (table will be in readonly mode): Replica /clickhouse/tables/clickhouse-main/02/insights.event_hourly_shard/replicas/example.com appears to be already active (host: example.com, port: 9009, tcp_port: 9000, database: insights, table: insights.event_hourly_shard, scheme: http, ). If you're sure it's not, try again in a minute or remove znode /clickhouse/tables/clickhouse-main/02/insights.event_hourly_shard/replicas/example.com/is_active manually. Code: 224, e.displayText() = DB::Exception: Replica /clickhouse/tables/clickhouse-main/02/insights.event_hourly_shard/replicas/example.com appears to be already active (host: example.com, port: 9009, tcp_port: 9000, database: insights, table: insights.event_hourly_shard, scheme: http, ). If you're sure it's not, try again in a minute or remove znode /clickhouse/tables/clickhouse-main/02/insights.event_hourly_shard/replicas/example.com/is_active manually, Stack trace (when copying this message, always include the lines below):
0. DB::Exception::Exception(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, int, bool) @ 0x87da03a in /usr/bin/clickhouse
1. DB::Exception::Exception<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >&>(int, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >&) @ 0x8860513 in /usr/bin/clickhouse
2. DB::ReplicatedMergeTreeRestartingThread::activateReplica() @ 0xfd39371 in /usr/bin/clickhouse
3. DB::ReplicatedMergeTreeRestartingThread::tryStartup() @ 0xfd37221 in /usr/bin/clickhouse
4. DB::ReplicatedMergeTreeRestartingThread::run() @ 0xfd360e5 in /usr/bin/clickhouse
5. DB::BackgroundSchedulePoolTaskInfo::execute() @ 0xefca140 in /usr/bin/clickhouse
6. DB::BackgroundSchedulePool::threadFunction() @ 0xefcc197 in /usr/bin/clickhouse
7. ? @ 0xefccf12 in /usr/bin/clickhouse
8. ThreadPoolImpl<std::__1::thread>::worker(std::__1::__list_iterator<std::__1::thread, void*>) @ 0x881a2df in /usr/bin/clickhouse
9. ? @ 0x881d803 in /usr/bin/clickhouse
10. start_thread @ 0x814a in /usr/lib64/libpthread-2.28.so
11. __clone @ 0xfcdc3 in /usr/lib64/libc-2.28.so
 (version 21.5.7.9 (official build))
```

**Comment by den-crane** [comment]
>Yes it's partitioned by event_at, but before inserting its grouped to 15 minutes.

You do understand that by design CH expects monthly partitioning? 
15 minutes partitions are unthinkable.

**Comment by mkocztorz** [comment]
Thanks for pointing it out. 
I did not find information that monthly partitions are expected. I've found that documentation warns about creating tables with over granular partitions (more than 1000) https://clickhouse.tech/docs/en/engines/table-engines/mergetree-family/custom-partitioning-key/#custom-partitioning-key. I'd be more than happy to read more about it if you have some links.
The schema I presented in this issue is work in progress, and it will not be final one as I already changed some things and I'm testing different approaches too. The event_shard table was from the beginning designed to store data for short period of time (although TTL is not included in the example, cause it's not related to the issue).
Could wrong approach to the partitioning be causing the problem with aggregated data?
